app/validators/declaration_date_validator.rb

```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+class DeclarationDateValidator < ActiveModel::Validator
+  RFC3339_DATE_REGEX = /\A\d{4}-\d{2}-\d{2}T(\d{2}):(\d{2}):(\d{2})([.,]\d+)?(Z|[+-](\d{2})(:?\d{2})?)?\z/i
+
+  def validate(record)
+    return if record.errors.any?
+
+    date_has_the_right_format(record)
+    declaration_within_schedule(record)
+  end
+
+private
+
+  def date_has_the_right_format(record)
+    return if record.raw_declaration_date.blank?
+    return if record.raw_declaration_date.match?(RFC3339_DATE_REGEX) && begin
+      Time.zone.parse(record.raw_declaration_date.to_s)
+    rescue ArgumentError
+      false
+    end
+
+    record.errors.add(:declaration_date, :invalid)
+  end
+
+  def declaration_within_schedule(record)
+    return unless record.schedule && record.declaration_date.present?
+
+    if record.declaration_date < record.schedule.applies_from.beginning_of_day
+      record.errors.add(:declaration_date, :declaration_before_milestone_start)
+    end
+
+    if record.schedule.applies_to.end_of_day <= record.declaration_date
```

Comment:
I think `milestone_date` in ECF doesn't equate to `applies_to` here as in ECF all milestone dates are `nil` for NPQ, and we only have payment date which we don't validate against in declarations, do you mind confirming @leandroalemao unless I'm missing something

Reply:
Ah yeah we're not migrating `milestone_date` over as it's `nil` for NPQ, we did a spike on that. The question is .. should we keep this & validate the the declaration against the payment_date (`applies_to`)? Or in another words.. is it ok to create a declaration after a schedule `applies_to` date?

Reply:
We have something similar here:
https://github.com/DFE-Digital/npq-registration/pull/1453/files#diff-92e39530416181739ced666c8a1c0e1e14f64d062ea6cabd5819387b76ef6f38R82

ECF had the check with `milestone_date`, but I removed it because we don't have `milestone_date` on NPQ.

Reply:
discussed in person: for now we're going to remove the validation on applies to